### ISSUE TITLE
Lua.GetFunction(string fullPath) changed behavior

### ISSUE THREAD

**Issue by theKBro** [NLua/NLua#346]
In earlier versions if the object at fullPath was nil GetFunction returned null.
Now if the object is not a LuaFunction a LuaFunction is created -> so GetFunction will never return null.

Is this really  the intended behavior?

**Comment by Valleriani** [comment]
This is also an issue I'm having. I have a dynamic loading of a LUA file. Some functions exist per file, some don't, depends on the 'monster' and if it's needed. If it didn't exist on load I'd set a flag and it'll ignore it, but now everything is returning a value and I'm not sure how to figure out that GetFunction has a actual function, that's valid. 

Obviously since this occurs, it think it has a function, runs the code, crashes.

**Comment by Valleriani** [comment]
I was able to fix it via doing

LuaFunction funcName = state["init"]

instead of using state.GetFunction. Above one returns null when needed. Dunno why GetFunction doesn't, but this works out fine and deals with returning the function fine.

**Comment by theKBro** [comment]
I'm using the same workaround.
Thanks for pointing out the new overhead, I had not checked this yet.

**Comment by Valleriani** [comment]
Yeah my game is pretty extensive on how it uses LUA for alot of things. Objects have their own state because LUA doesn't seem to play nice with one state if you need a similar function name, though I could call it attack_mobname(), attack_mobname2() etc. Might not be the best solution, might be alright, but it was mangable before. 120MB isn't too bad. 500MB isn't too bad, I could live with it, but for some reason to load up the game takes far longer with this new update + 5.3. I reverted back to the original NLUA with 5.2 for now because of it, as it worked fine and was ok as it is.  This is my work machine, which is a lot faster:

![Untitled](https://user-images.githubusercontent.com/503080/72748771-77506480-3bb8-11ea-94ce-548742b5cfda.png)

3x longer, much more ram, nothing changed, updated LUA and adjusted the function name issue above. Thing is, with my game sometimes I'm initlizing new objects after run time as well, init does a heavy load but if it slower, for sure it's going to be slower in game. For me, I did really want LUA 5.3 for the long integers, but it's so much slower and I don't understand why.

**Comment by Valleriani** [comment]
Will state this though. I register a bunch of functions to each state, the LoadState() function seems to be a big deal with this, aka RegisterFunction()..

On the old it's an average of 00:00:00.0590169 per state. 

The new NLUA is 00:00:00.4569571 per state. 

I guess the easiest way to replicate this would be to add a bunch of RegisteredFunctions and load them in a cycle, they're a lot slower and use way more memory.

**Comment by theKBro** [comment]
any news here?

**Comment by albertotirla** [comment]
Hy there!  
Something like the problem mentioned above begun to plague me too, when doing some heavier stuff in my app. In searching for a solution, I found this issue.

So, from what I gathered, It does seemn that the repo is abandoned or something.  
I know the person maintaining this thing does it voluntarely, on their own, and, I suppose by how slow are things going, in their free time.  
However, I find it a bit odd that there are issues that got no answer from the developer for a...let's say...long time indeed. For example, I got a problem before, it's rather unexpected, but here we go...  
I'm refering to #403 , btw.

**Comment by Valleriani** [comment]
Yeah it is a shame. At the moment however I find the NLUA with 5.2 works fine. I really wanted to get to 5.3 but with the problems here including the GetFunction issues, memory usage being super high, and simply it's just far slower, my next upgrade is probably going to be off NLUA if I ever do it. It's just a hard thing to even fathom because my game has been built up and we have several scripts now (~800), and the old version is somewhat OK. 

If you use NUGET I believe the package version for the older one is 1.3.2.1 for 5.2 . I would attempt to try that one for now and ignore the current 1.4x or 1.5x versions until the developer looks into it. You lose out on some things like long integers and some new LUA functions though but for me it's not a huge issue at this time.

**Comment by viniciusjarina** [comment]
Hi @Valleriani thank you for the detailed report, have you chacked which changes caused the memory consumption to increase ?

Do you happen to have a small test that could repro and isolate the problem?

NLua is quite huge, saying "the new version consumes more memory" make really hard to address the problem

@albertotirla take a look on the tests to see how to deal with static type, there are several ways to do it. I know NLua lacks a good documentation, any contribution and feedback will be welcome.

**Comment by EvaisaGiac** [comment]
I can confirm, I just downgraded to the 5.2 version because my game's framerate got cut in half by lua function calls in the newer versions of NLua.

**Comment by viniciusjarina** [comment]
@EvaisaGiac do you have more info where exactly we did regressed? Did you  profile the function?

**Comment by EvaisaGiac** [comment]
> @EvaisaGiac do you have more info where exactly we did regressed? Did you profile the function?

It seems to specifically be calling LuaFunctions that seems to cause the performance issues.
I am calling a Update function every frame, which is fine in the 5.2 version, but tanks the performance massively in any version after that.
I did not profile anything, I have never profiled functions in C# before so I am not quite sure where I would start.
The code causing issues is simply:
```cs        
        public void Update(GameTime gameTime)
        {
            try
            {
                if (_updateFunc != null)
                {
                    _updateFunc.Call();
                }
            }
            catch (Exception e) { }
        }
```

**Comment by EvaisaGiac** [comment]
Is there any way to get the source code of 1.3.2.1? I want to compare them to see if i can figure out where the issue lies.

**Comment by viniciusjarina** [comment]
1.3.2 is about 8 years old. You can compare using git 

https://github.com/NLua/NLua/compare/26031db68cadceb1325e2ba2667d18719147264a..44b9cd61a9434aa6f46526c9587c364fa10c6530

**Comment by EvaisaGiac** [comment]
Okay the issue is very strange, I profiled the Lua function call, and the performance of 1.6.0 is roughly equal, if not better than 1.3.2.1, however it still halfs the framerate of my game.
So I still do not know what is causing the performance drop.

Here are some profiling results:

1.3.2.1:
```
[2022/08/10 02:34:55.716] Elapsed Time: 0.0171
[2022/08/10 02:34:55.716] Elapsed Time: 0.2032
[2022/08/10 02:34:55.732] Elapsed Time: 0.0016
[2022/08/10 02:34:55.732] Elapsed Time: 0.0386
[2022/08/10 02:34:55.749] Elapsed Time: 0.0067
[2022/08/10 02:34:55.749] Elapsed Time: 0.0729
[2022/08/10 02:34:55.766] Elapsed Time: 0.0061
[2022/08/10 02:34:55.766] Elapsed Time: 0.1068
[2022/08/10 02:34:55.782] Elapsed Time: 0.0021
[2022/08/10 02:34:55.782] Elapsed Time: 0.0634
```

1.6.0
```
[2022/08/10 02:31:29.461] Elapsed Time: 0.0077
[2022/08/10 02:31:29.461] Elapsed Time: 0.0987
[2022/08/10 02:31:29.461] Elapsed Time: 0.0006
[2022/08/10 02:31:29.461] Elapsed Time: 0.0263
[2022/08/10 02:31:29.461] Elapsed Time: 0.0005
[2022/08/10 02:31:29.461] Elapsed Time: 0.036
[2022/08/10 02:31:29.461] Elapsed Time: 0.0005
[2022/08/10 02:31:29.461] Elapsed Time: 0.0242
[2022/08/10 02:31:29.461] Elapsed Time: 0.0004
[2022/08/10 02:31:29.461] Elapsed Time: 0.0243
```

I do not have any other NLua functions that run every frame as far as I know which makes the consistent frame tank in newer versions even more confusing.
